Create node PortGroup and add pyright port

gpanders opened the pull request:
#### Description

<!-- Note: it is best to make pull requests from a branch rather than from master -->
Add a new node PortGroup for packages that are hosted on the npmjs registry.
These are packages that can typically be installed with `npm install -g`. The
PortGroup essentially mimics this behavior (downloads the .tgz file from the
registry and `npm installs` it).

I developed this while working on a port for pyright, which is also included.

Note this is different than the npm PortGroup proposed in https://github.com/macports/macports-ports/pull/11668. That PR proposed a group just for `npm*` ports, while this PortGroup is designed for packages that are hosted on the npmjs registry (such as pyright).

###### Type(s)
<!-- update (title contains ": U(u)pdate to"), submission (new Portfile) and CVE Identifiers are auto-detected, replace [ ] with [x] to select -->

- [ ] bugfix
- [x] enhancement
- [ ] security fix

###### Tested on
<!-- Triple-click and copy the next line and paste it into your shell. It will copy your OS and Xcode version to the clipboard. Paste it here replacing this section.
sh -c 'printf "%s\n" "macOS `sw_vers -productVersion` `sw_vers -buildVersion` `uname -m`" "`xcodebuild -version|awk '\''NR==1{x=$0}END{print x" "$NF}'\''`"'|tee /dev/tty|pbcopy
-->
macOS 12.0.1 21A559 arm64
Xcode 13.1 13A1030d

###### Verification <!-- (delete not applicable items) -->
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [x] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL? <!-- Please don't open a new Trac ticket if you are submitting a pull request. -->
- [x] checked your Portfile with `port lint`?
- [x] tried existing tests with `sudo port test`?
- [x] tried a full install with `sudo port -vst install`?
- [x] tested basic functionality of all binary files?

<!-- Use "skip notification" (surrounded with []) to avoid notifying maintainers -->

mascguy commented on devel/pyright/Portfile:
```diff
@@ -0,0 +1,22 @@
+PortSystem          1.0
+PortGroup           node 1.0
+
+name                pyright
```
@reneeotten Any concerns relative to the port name? While it matches the upstream project, the `py` prefix might make folks think it's a Python-based port.

I'm almost wondering if it should be named `node-pyright`?

But if you're fine with the name as-is, then I'm good too!  ;-)

gpanders replied:
My 2¢: the fact that it’s written in Node is an implementation detail and not something users should know or care about, so I’d vote for just keeping “pyright”. Besides, if MS decides to one day RIIR we wouldn’t need to rename it :)

mascguy replied:
> My 2¢: the fact that it’s written in Node is an implementation detail and not something users should know or care about, so I’d vote for just keeping “pyright”. Besides, if MS decides to one day RIIR we wouldn’t need to rename it :)

Yep, and so long as others are fine with it, I'm good too. But just wanted to bring it up, to see if anyone feels differently.

reneeotten replied:
I think we should follow the upstream name; Python ports are name `py-<name>` so I don't think there is any confusion here.

mascguy replied:
> I think we should follow the upstream name; Python ports are name `py-<name>` so I don't think there is any confusion here.

Sounds good to me. Thanks for the feedback Renee!

mascguy commented:
Overall though, I like the idea of a `node` portgroup. Great stuff!

mascguy commented:
LGTM. @ci42?

reneeotten commented:
how many ports would potentially make use of this PortGroup? There isn't all that much code in it so if there are only a handful of ports I am not really convinced that it's necessary to introduce a PortGroup for it. But if people agree it's a good idea it should support the typical things, including a `livecheck`, which I don't see right now but admittedly I haven't tried anything with this PR locally...

gpanders commented:
> how many ports would potentially make use of this PortGroup?

Any package that can be installed using `npm install -g`. There aren't many in MacPorts right now, but some examples I can think of off the top of my head:

* `bitwarden-cli`
* [`prettier`](https://prettier.io/)
* `eslint`

I can add the `livecheck`. Anything else you'd like to see included?

gpanders commented:
@reneeotten Anything else I can do to help move this along?

mascguy commented:
> @reneeotten Anything else I can do to help move this along?

LGTM.

@reneeotten and @ci42 ?

gpanders commented:
Ping.

I'd like to get pyright added -- while I think the node port group is useful and I would like to see that merged, if it's problematic I can redo the PR to omit that and only include the pyright port.

reneeotten commented:
I personally don't think there is a need for a PortGroup here, but gain that's just my opinion (you could ask on the MacPorts' devel list to solicit opinions). The `pyright` port seems like a nice addition and could be merged quickly; finally, I am not familiar at all with `npm` so will myself not merge the change in the `npm8` Portfile without @ci42's approval.

So practically, if you want me to merge something I'd be happy to do that for a PR with only the `pyright` port.

mascguy commented:
I'm fine with the new portgroup, given that it can significantly simplify future ports.

My only thought is that it might be better to name it `npm`.

gpanders commented:
As requested, I renamed the PortGroup to `npm`.

I also removed myself as a maintainer from the `pyright` port as I'm no longer using MacPorts day to day.

FranklinYu commented:
I feel that Prettier and ESLint seldom got installed globally. Most projects put these tools into their dev-dependencies.

reneeotten commented:
> As requested, I renamed the PortGroup to `npm`.
> 
> I also removed myself as a maintainer from the `pyright` port as I'm no longer using MacPorts day to day.

in that case, let's just only add the `pyright` port and not the PortGroup...

reneeotten commented:
> I feel that Prettier and ESLint seldom got installed globally. Most projects put these tools into their dev-dependencies.

I am not sure I understand your comment @FranklinYu - what does it have to do with this PR? Perhaps I'm just missing something as I admittedly have no knowledge whatsoever about `npm`, `nodejs` and such...

reneeotten commented:
@blair do you have any comments/opinion on this PR (I don't use `npm` or `nodejs`...). For example, should we be moving to `nodejs17` as a dependency for `npm8` or not, is a separate PG useful here, etcetera?

mascguy commented:
> @blair do you have any comments/opinion on this PR (I don't use `npm` or `nodejs`...). For example, should we be moving to `nodejs17` as a dependency for `npm8` or not, is a separate PG useful here, etcetera?

I can confirm that we definitely do NOT want to move to Node 17, as there are some critical issues related to OpenSSL 3. Here's one example that explains it, but this isn't an isolated issue:

https://itsmycode.com/error-digital-envelope-routines-unsupported/

In short, many developers and users are simply staying with Node 16 at present. Plus Node 17 isn't an LTS release, so that's another reason not to use that as the default.

FranklinYu commented:
@reneeotten My comment was to reply https://github.com/macports/macports-ports/pull/13119#issuecomment-980783302.
